wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/validator/DevfileIntegrityValidator.java

```diff
@@ -111,20 +113,36 @@ public void validateContentReferences(Devfile devfile, FileContentProvider provi
   }
 
   private Set<String> validateComponents(Devfile devfile) throws DevfileFormatException {
-    Set<String> existingNames = new HashSet<>();
+    Set<String> definedAliases = new HashSet<>();
     Component editorComponent = null;
+
+    Map<String, Set<String>> idsPerComponentType = new HashMap<>();
+
     for (Component component : devfile.getComponents()) {
-      if (!existingNames.add(component.getName())) {
+      if (component.getAlias() != null && !definedAliases.add(component.getAlias())) {
         throw new DevfileFormatException(
-            format("Duplicate component name found:'%s'", component.getName()));
+            format("Duplicate component alias found:'%s'", component.getAlias()));
       }
+
+      if (!idsPerComponentType
+          .computeIfAbsent(component.getType(), __ -> new HashSet<>())
+          .add(getIdentifiableComponentName(component))) {
```

Comment:
This check ensures only that `IdentifiableComponentName` is unique per component type, but not along all components types. See the following tests cases:
```java
  @Test(
      expectedExceptions = DevfileFormatException.class)
  public void shouldThrowAnExceptionIfMultipleIdentifableComponentsNamesAreNotUnique() throws Exception {
    // given
    DevfileImpl devfile = new DevfileImpl(initialDevfile);

    ComponentImpl k8s = new ComponentImpl();
    k8s.setType(KUBERNETES_COMPONENT_TYPE);
    k8s.setReference("k8s.yaml");

    ComponentImpl os = new ComponentImpl();
    os.setType(OPENSHIFT_COMPONENT_TYPE);
    os.setReference("k8s.yaml");

    devfile.getComponents().add(k8s);
    devfile.getComponents().add(os);

    // when
    integrityValidator.validateDevfile(devfile);

    // then
    // exception is thrown
  }

  @Test(
      expectedExceptions = DevfileFormatException.class)
  public void shouldThrowAnExceptionIfMultipleIdentifableComponentsNamesAreNotUnique2() throws Exception {
    // given
    DevfileImpl devfile = new DevfileImpl(initialDevfile);

    ComponentImpl tool1 = new ComponentImpl();
    tool1.setType(EDITOR_COMPONENT_TYPE);
    tool1.setId("id");

    ComponentImpl tool2 = new ComponentImpl();
    tool2.setType(PLUGIN_COMPONENT_TYPE);
    tool2.setReference("id");

    devfile.getComponents().add(tool1);
    devfile.getComponents().add(tool2);

    // when
    integrityValidator.validateDevfile(devfile);

    // then
    // exception is thrown
  }
```
I think that cases are kind of theoretical corner and not sure if we need to handle it, but at least it makes sense to make it clear that `Components.getIdentifiableComponentName` returns values that is not unique withing all components

Reply:
I don't think this actually makes any difference.

We only need any kind of identifier of the components for a) error reporting and b) referencing them within the devfile.

For error reporting, we now always use a combo "identifiable name" + "type" (as of my latest commit) which makes the components in your test cases distinguishable for the user.

For referencing the components we already require the alias, which is required to be unique in the devfile as a whole.

Reply:
> we now always use a combo "identifiable name" + "type" 

Maybe it makes sense to add a note to java doc of `Components.getIdentifiableComponentName` to make it clear that IdentifiableComponentName can be used to identify component only along with its type?
Now it just says

> Get a name that can be used to identify the component in the devfile.

Reply:
I added more info to the javadoc.